tests/storages_tests/test_with_server.py

```diff
@@ -115,7 +118,17 @@ def test_loaded_trials(storage_url: str) -> None:
 )
 def test_store_infinite_values(input: float, expected: float, storage_url: str) -> None:
 
-    storage = optuna.storages.RDBStorage(url=storage_url)
+    storage: Union[RDBStorage, RedisStorage]
+    # This way is a workaround. When optuna.storages.get_storage() supports raw
+    # storage mode, use it like this.
+    # storage = optuna.storages.get_storage(url=storage_url, raw=True)
+    if storage_url.startswith("redis"):
+        # Only RDB can convert max/min value to inf.
+        if input != float("inf") and input != -float("inf"):
+            return
```

Comment:
If we have a tracking issue about supporting `float(inf)`, could you add a link for the developers? In addition, how about replacing this silent return with `pytest.skip("RedisStorage does not accept float('inf')")` ?

Reply:
I have confirmed that the current Redis storage already supports `inf` to store it in the storage. 

<details>
<summary>Test for study.optimize</summary>

```python
import optuna
import numpy as np

def objective(trial):
    return float(np.inf)

# In-memory storage
study = optuna.create_study()
study.optimize(objective, n_trials=1)
print(f"In-memory : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (SQLite)
study = optuna.create_study(storage='sqlite:///../sqlite.db')
study.optimize(objective, n_trials=1)
print(f"SQLite    : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (MySQL)
study = optuna.create_study(storage="mysql+pymysql://root:test@localhost/mysql")
try:
    study.optimize(objective, n_trials=1)
except Exception as e:
    print(e)
print(f"MySQL     : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (PostgreSQL)
study = optuna.create_study(storage="postgresql+psycopg2://postgres:test@localhost:15432/postgres")
study.optimize(objective, n_trials=1)
print(f"PostgreSQL: Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# Redis storage
study = optuna.create_study(storage="redis://localhost:6379/0")
try:
    study.optimize(objective, n_trials=1)
except Exception as e:
    print(e)
finally:
    print(f"Redis     : Trial state is {study.trials[-1].state}")
print(study.trials[-1])
```

```
(venv) mamu@HideakinoMacBook-puro optuna % python inf-objective.py
[I 2022-02-03 16:38:09,480] A new study created in memory with name: no-name-5b41f645-2d08-4af3-b343-b8cfd447e88e
[I 2022-02-03 16:38:09,481] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
In-memory : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 38, 9, 481592), datetime_complete=datetime.datetime(2022, 2, 3, 16, 38, 9, 481728), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=0, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:38:09,548] A new study created in RDB with name: no-name-d3c8c30d-ed7e-4d8e-82b9-fa0ddd75127f
[I 2022-02-03 16:38:09,611] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
SQLite    : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 38, 9, 569079), datetime_complete=datetime.datetime(2022, 2, 3, 16, 38, 9, 582011), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=24, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:38:09,742] A new study created in RDB with name: no-name-f8300a3d-90c1-4909-bf2a-3c7ec1042da1
[I 2022-02-03 16:38:09,940] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
MySQL     : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 38, 9, 793344), datetime_complete=datetime.datetime(2022, 2, 3, 16, 38, 10), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=48, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:38:10,171] A new study created in RDB with name: no-name-eae39a96-126f-4af9-a43e-5f4d1def5146
[I 2022-02-03 16:38:10,317] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
PostgreSQL: Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 38, 10, 215340), datetime_complete=datetime.datetime(2022, 2, 3, 16, 38, 10, 245949), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=27, state=TrialState.COMPLETE, value=None)
/Users/mamu/Dev/pfn/optuna/optuna/storages/__init__.py:29: ExperimentalWarning: RedisStorage is experimental (supported from v1.4.0). The interface can change in the future.
  return _CachedStorage(RedisStorage(storage))
[I 2022-02-03 16:38:10,336] A new study created in Redis with name: no-name-0000000016
[I 2022-02-03 16:38:10,343] Trial 0 finished with value: inf and parameters: {}. Best is trial 0 with value: inf.
Redis     : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 38, 10, 338778), datetime_complete=datetime.datetime(2022, 2, 3, 16, 38, 10, 340677), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=16, state=TrialState.COMPLETE, value=None)
```

</details>

<details>
<summary>Test for trial.report</summary>

```python
import optuna
import numpy as np

def objective(trial):
    trial.report(float(np.inf), 1)
    return 1

# In-memory storage
study = optuna.create_study()
study.optimize(objective, n_trials=1)
print(f"In-memory : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (SQLite)
study = optuna.create_study(storage='sqlite:///../sqlite.db')
study.optimize(objective, n_trials=1)
print(f"SQLite    : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (MySQL)
study = optuna.create_study(storage="mysql+pymysql://root:test@localhost/mysql")
try:
    study.optimize(objective, n_trials=1)
except Exception as e:
    print(e)
print(f"MySQL     : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (PostgreSQL)
study = optuna.create_study(storage="postgresql+psycopg2://postgres:test@localhost:15432/postgres")
study.optimize(objective, n_trials=1)
print(f"PostgreSQL: Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# Redis storage
study = optuna.create_study(storage="redis://localhost:6379/0")
try:
    study.optimize(objective, n_trials=1)
except Exception as e:
    print(e)
finally:
    print(f"Redis     : Trial state is {study.trials[-1].state}")
print(study.trials[-1])
```

```
venv) mamu@HideakinoMacBook-puro 3205-3206 % python inf-report.py
[I 2022-02-03 16:39:44,183] A new study created in memory with name: no-name-55b7342a-1401-4c04-a5b0-b3ef6ec14c84
[I 2022-02-03 16:39:44,185] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
In-memory : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[1.0], datetime_start=datetime.datetime(2022, 2, 3, 16, 39, 44, 184687), datetime_complete=datetime.datetime(2022, 2, 3, 16, 39, 44, 184845), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={1: inf}, trial_id=0, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:39:44,274] A new study created in RDB with name: no-name-3ce44212-c256-42d1-914e-ba8ab64bd5a0
[I 2022-02-03 16:39:44,349] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
SQLite    : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[1.0], datetime_start=datetime.datetime(2022, 2, 3, 16, 39, 44, 295707), datetime_complete=datetime.datetime(2022, 2, 3, 16, 39, 44, 320515), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={1: inf}, trial_id=29, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:39:44,482] A new study created in RDB with name: no-name-93b70a4e-4a32-475d-bfbc-ac071b8fbbc4
[I 2022-02-03 16:39:44,706] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
MySQL     : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[1.0], datetime_start=datetime.datetime(2022, 2, 3, 16, 39, 44, 544455), datetime_complete=datetime.datetime(2022, 2, 3, 16, 39, 45), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={1: inf}, trial_id=49, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:39:44,935] A new study created in RDB with name: no-name-7e646d7c-53b3-4e67-98cc-dc2b3c2c194d
[I 2022-02-03 16:39:45,083] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
PostgreSQL: Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[1.0], datetime_start=datetime.datetime(2022, 2, 3, 16, 39, 44, 979182), datetime_complete=datetime.datetime(2022, 2, 3, 16, 39, 45, 19134), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={1: inf}, trial_id=28, state=TrialState.COMPLETE, value=None)
/Users/mamu/Dev/pfn/optuna/optuna/storages/__init__.py:29: ExperimentalWarning: RedisStorage is experimental (supported from v1.4.0). The interface can change in the future.
  return _CachedStorage(RedisStorage(storage))
[I 2022-02-03 16:39:45,100] A new study created in Redis with name: no-name-0000000017
[I 2022-02-03 16:39:45,107] Trial 0 finished with value: 1.0 and parameters: {}. Best is trial 0 with value: 1.0.
Redis     : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[1.0], datetime_start=datetime.datetime(2022, 2, 3, 16, 39, 45, 102660), datetime_complete=datetime.datetime(2022, 2, 3, 16, 39, 45, 104854), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={1: inf}, trial_id=17, state=TrialState.COMPLETE, value=None)
```

</details>

<details>
<summary>Test for study.tell</summary>

```python
import optuna
import numpy as np


# In-memory storage
study = optuna.create_study()
trial = study.ask()
study.tell(trial, float(np.inf))
print(f"In-memory : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (SQLite)
study = optuna.create_study(storage='sqlite:///../sqlite.db')
trial = study.ask()
study.tell(trial, float(np.inf))
print(f"SQLite    : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (MySQL)
study = optuna.create_study(storage="mysql+pymysql://root:test@localhost/mysql")
trial = study.ask()
try:
    study.tell(trial, float(np.inf))
except Exception as e:
    print(e)
print(f"MySQL     : Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# RDB storage (PostgreSQL)
study = optuna.create_study(storage="postgresql+psycopg2://postgres:test@localhost:15432/postgres")
trial = study.ask()
study.tell(trial, float(np.inf))
print(f"PostgreSQL: Trial state is {study.trials[-1].state}")
print(study.trials[-1])

# Redis storage
study = optuna.create_study(storage="redis://localhost:6379/0")
trial = study.ask()
try:
    study.tell(trial, float(np.inf))
except Exception as e:
    print(e)
finally:
    print(f"Redis     : Trial state is {study.trials[-1].state}")
print(study.trials[-1])
```

```
(venv) mamu@HideakinoMacBook-puro 3205-3206 % python inf-tell.py 
[I 2022-02-03 16:42:16,267] A new study created in memory with name: no-name-bfd972f3-6428-4f86-8781-269269a25e0b
In-memory : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 42, 16, 267776), datetime_complete=datetime.datetime(2022, 2, 3, 16, 42, 16, 267909), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=0, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:42:16,336] A new study created in RDB with name: no-name-2919cfc3-61c8-46c6-bc1f-7b5884d36e2c
SQLite    : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 42, 16, 354312), datetime_complete=datetime.datetime(2022, 2, 3, 16, 42, 16, 367098), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=30, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:42:16,511] A new study created in RDB with name: no-name-6e099bcf-cdea-457f-bfb3-2e3f56b6cb9c
MySQL     : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 42, 16, 565982), datetime_complete=datetime.datetime(2022, 2, 3, 16, 42, 17), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=50, state=TrialState.COMPLETE, value=None)
[I 2022-02-03 16:42:16,868] A new study created in RDB with name: no-name-15f997d7-266d-40ec-8d64-51ad95796f27
PostgreSQL: Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 42, 16, 911200), datetime_complete=datetime.datetime(2022, 2, 3, 16, 42, 16, 936047), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=29, state=TrialState.COMPLETE, value=None)
/Users/mamu/Dev/pfn/optuna/optuna/storages/__init__.py:29: ExperimentalWarning: RedisStorage is experimental (supported from v1.4.0). The interface can change in the future.
  return _CachedStorage(RedisStorage(storage))
[I 2022-02-03 16:42:16,974] A new study created in Redis with name: no-name-0000000018
Redis     : Trial state is TrialState.COMPLETE
FrozenTrial(number=0, values=[inf], datetime_start=datetime.datetime(2022, 2, 3, 16, 42, 16, 976742), datetime_complete=datetime.datetime(2022, 2, 3, 16, 42, 16, 978574), params={}, distributions={}, user_attrs={}, system_attrs={}, intermediate_values={}, trial_id=18, state=TrialState.COMPLETE, value=None)
```

</details>

The test of `test_store_infinite_values` was created for only RDB, which automatically converts the value greater than `np.finfo(np.float32).max` into `float("inf")`. Redis storage does not have this specification. Therefore, I guess, @masap skip the test if the `input` is not `+/- float("inf")`.

Reply:
We introduced a specification in #3238 that converts values larger than `np.finfo(np.float32).max` to `float("inf")` in the RDB.

Reply:
> Therefore, I guess, @masap skip the test if the input is not +/- float("inf").

My code is doing it...

float("inf") -> doing test
-float("inf") -> doint test
other -> skip test

Reply:
> In addition, how about replacing this silent return with pytest.skip("RedisStorage does not accept float('inf')") ?

I will fix to return "RedisStorage does not convert max/min to inf/-inf".

Reply:
> Therefore, I guess, @masap skip the test if the input is not +/- float("inf").

Yes. (I removed previous my comment because I mis-understood your comment)

> We introduced a specification in https://github.com/optuna/optuna/pull/3238 that converts values larger than np.finfo(np.float32).max to float("inf") in the RDB.

Yes. I will add the link to it.

Reply:
Fixed.